### PR TITLE
feat(client): account for portfolio project without image

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

These changes aim to make visually pleasing to have a project without image, I haven't checked the mobile layout yet

image

While reviewing this, to understand `:has` think of it as an `if` statement.

<pre>
if (`parent` has `something`) {
display these properties
}
</pre>

### CODE REVIEW

File: client/src/components/profile/components/portfolio.css [hunk 1]
```diff
@@ -147,3 +165,45 @@
...
+      grid-column: 2;
+      grid-row: 2;
+    }
+
+    .portfolio-container a {
+      grid-column: 2;
+      grid-row: 3;
+    }
+  }
+}
```
Comment (Sboonny) [inline client/src/components/profile/components/portfolio.css:209]: `has` selector isn't supported in Firefox, so we are using `selector` function that isn't supported in opera to tell the browser to display the old style

**Comment by bbsmooth** [comment]
> understand :has think of it as an if statement

Just a reminder that Firefox does not support `:has` yet.

**Review by raisedadead — DISMISSED** [review]
> Just a reminder that Firefox does not support `:has` yet.

Makes sense the tests are failing.

**Review by raisedadead — CHANGES_REQUESTED** [review]
Does this mean it will behave in different browsers? 

If yes, what is the justification here? How does this feature get managed long-term? Do we need to record the differences somewhere as docs so that it is not forgotten?

**Comment by Sboonny** [comment]
> Does this mean it will behave in different browsers?

This `@supports not (selector(:has))` take care of card looking broken in different browser.

> what is the justification here?

Go to `/setting` and create a portfolio project without adding image, check it in the profile, even without you adding image, you will see broken image, this sort that issue.

> How does this feature get managed long-term?

In long term, Firefox supports `has` and `@supports not (selector(:has))` becomes redundant, or it acts the same as the current PR.

> Do we need to record the differences somewhere as docs so that it is not forgotten?

I wouldn't.

**Comment by raisedadead** [comment]
I see where you are coming from. But IMHO, other UX considerations need to be made. 


This is not (and should not be) a quick CSS fix. We should have different layouts (tiny functions that render a certain container in the `portfolio-projects` components) for the cards:

1. When only a title and a URL are available, aka the basics.
2. A description is also available in addition to the basics above.
3. An image is also available in addition to the basics above.
4. Descriptions and images are also available in addition to the basics.

So each condition above get's its own "container" card. Then you have a simple CSS style for each card. Then you can render the container based on the attributes you get in your props.

That should be the logical approach here.

**Comment by Sboonny** [comment]
> I see where you are coming from. But IMHO, other UX considerations need to be made.
> 
> This is not (and should not be) a quick CSS fix. We should have different layouts (tiny functions that render a certain container in the `portfolio-projects` components) for the cards:
> 
> 1. When only a title and a URL are available, aka the basics.
> 2. A description is also available in addition to the basics above.
> 3. An image is also available in addition to the basics above.
> 4. Descriptions and images are also available in addition to the basics.
> 
> So each condition above get's its own "container" card. Then you have a simple CSS style for each card. Then you can render the container based on the attributes you get in your props.
> 
> That should be the logical approach here.

I don't like the idea of creating four components and four styles, for single card.

So let's do it like this, I will create a PR for validation the input, and include a fallback image, if the string is empty. maybe the freeCodeCamp Glyph.

This way, we don't have to use `has` or create multiple components to account for different states.

**Comment by naomi-lgbt** [comment]
> I don't like the idea of creating four components and four styles, for single card.

Why would that need to be four components? Should be something for conditional rendering, no?

**Comment by Sboonny** [comment]
you forgot the to ask about the four styles 🙃, the number doesn't matter. IT will become redundant later, if we add fallback image, so I prefer adding the fallback image.

**Comment by raisedadead** [comment]
I advise discussing the possible solutions before jumping to solutions (creating PRs etc.). It wastes your time and, more importantly, the person looking at your work.

What you are proposing seems like a band-aid to avoid a UX change. That is, it may not look aesthetically pleasing, but sure, it is less code. That's not how it should be – we are in the business of writing code.

So going back to a more scientific discussion, if you will. 

Could you elaborate more on this first:

"I don't like the idea of creating four components and four styles for a single card."

We want strong reasons to move on to the next UX idea - not just a personal preference. And to be clear, these may not be four components but conditional ways of rendering information.

Thanks.

**Comment by Sboonny** [comment]
> "I don't like the idea of creating four components and four styles for a single card."

Sorry, a better wording can be, I like the idea of fallback image more, because the images can be set by the campers as default fallback to their projects, which can be their logo or picture of their selves, which can contribute to personalizing the profile page.


Although changing the card layout is one of the plan too. Although possible, I don't think it's worth it to put in the effort to make a different condition and layout, because we don't have the tools currently to make developing this as simple as the other solution. 

The condition rendering solution for the card can be different in the future, while the fallback image logic should still be the same.

**Comment by Sboonny** [comment]
Do I start working on this PR again? or is there something that we need to address before I do so?

**Pull request closed without merging**